cpp/src/arrow/flight/sql/server_session_middleware_factory.h

```diff
@@ -0,0 +1,57 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// ServerSessionMiddlewareFactory, factored into a separate header for testability
+
+#pragma once
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <shared_mutex>
+#include <utility>
+#include <vector>
+
+#include <arrow/flight/sql/server_session_middleware.h>
+
+namespace arrow {
+namespace flight {
+namespace sql {
+
+/// \brief A factory for ServerSessionMiddleware, itself storing session data.
+class ServerSessionMiddlewareFactory : public ServerMiddlewareFactory {
```

Comment:
You might need this

```suggestion
class ARROW_FLIGHT_SQL_EXPORT ServerSessionMiddlewareFactory : public ServerMiddlewareFactory {
```

Reply:
Thanks, hadn't even thought about that for something not exported for developer use....

Ran into another one, not a whole lot going on in the diff between this `cpp/src/arrow/flight/sql/CMakeLists.txt` and `origin/main`'s but I presume this is not failing on `main`:  https://github.com/indigophox/arrow/actions/runs/7600320482/job/20698418028  — thoughts?